sm_ipt/sm_ipt.c

```diff
@@ -0,0 +1,450 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+
+#include <stdint.h>
+#include <stddef.h>
+#include <stdbool.h>
+
+#include "sm_ipt.h"
+#include "sm_ipt_log.h"
+#include "sm_ipt_errno.h"
+
+
+#define FLAG_RELEASE 0x80
+
+#define WORD_SIZE sizeof(uint32_t)
+#define WORD_BITS (8 * sizeof(uint32_t))
+
+#if defined(CONFIG_SM_IPT_NUM_BLOCKS_32)
+typedef uint32_t mask_t;
+typedef int32_t smask_t;
+#define NUM_BLOCKS 32
+#define mask_clz sm_ipt_os_clz32
+#elif defined(CONFIG_SM_IPT_NUM_BLOCKS_64)
+typedef uint64_t mask_t;
+typedef int64_t smask_t;
+#define NUM_BLOCKS 64
+#define mask_clz sm_ipt_os_clz64
+#else
+#error Number of shared dynamic memory blocks is not configured.
+#endif
+
+#define ALLOCABLE_MULTIPLY (WORD_SIZE * NUM_BLOCKS)
+#define QUEUE_INDEX_MASK (NUM_BLOCKS - 1)
+
+/* Worst case scenario: NUM_BLOCKS count of messages waiting to be sent and NUM_BLOCKS count of
+ * buffers waiting to be freed + 1 item used as indicator of full queue
+ */
+#define QUEUE_ITEMS_CNT (2 * NUM_BLOCKS + 1)
+
+#define memory_corrupted_error() sm_ipt_os_fatal(); \
+	SM_IPT_ERR("Shared memory corrupted"); \
+	SM_IPT_ASSERT(0)
+
+
+struct sm_ipt_shm_queue {
+	uint32_t queue_tx;
+	uint32_t queue_rx;
+	uint8_t queue[QUEUE_ITEMS_CNT];
+	uint8_t handshake;
+	uint16_t _padding; /* Care should be taken to align allocable to word boundary */
```

Comment:
I might be wrong but wouldn't this padding will be added automatically by compiler ?

Reply:
Not necessarily. In this case uint8_t allocable needs to be aligned to word boundary. Allocable is of type uint8_t so compilers will not try to align it to word boundary.

Reply:
This is GCC specific, but we want to keep the code compiler-independent.